File: my_py_pkg/my_py_pkg/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int64

# Inherit from Node obj
class NumberPublisherNode(Node):
    def __init__(self):
        super().__init__("number_publisher")

        # declare parameter and set its default value. I can change that
        # from the terminal when launching the node:
        # ros2 run my_py_pkg number_publisher --ros-args -p number_to_publish:=5
        self.declare_parameter("number_to_publish", 4)
        self.declare_parameter("publish_frequency", 1.0)

        self.number_ = self.get_parameter("number_to_publish").value
        self.publish_frequency_ = self.get_parameter("publish_frequency").value

        # create_publisher(data_type_of_topic, topic_name, queue_size)
        self.publisher_ = self.create_publisher(Int64, "number", 10)
        self.timer_ = self.create_timer(1.0 / self.publish_frequency_, self.publish_number)
        # For debugging
        self.get_logger().info("Number Publisher has been started.")

    def publish_number(self):
        num = Int64()
        num.data = self.number_
        self.publisher_.publish(num)

def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisherNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__=="__main__":
    main()

File: my_cpp_pkg/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int64.hpp"

class NumberPublisherNode : public rclcpp::Node
{
public:
    NumberPublisherNode() : Node("number_publisher")
    {   
        this->declare_parameter("number_to_publish", 4);
        this->declare_parameter("publish_frequency", 1.0);

        /* In cpp I have also to cast the param type in order to match the declared
        type of the variable that is getting assigned */
        number_ = this->get_parameter("number_to_publish").as_int();
        double publish_frequency_ = this->get_parameter("publish_frequency").as_double();

        publisher_ = this->create_publisher<std_msgs::msg::Int64>("number", 10);
        timer_ = this->create_wall_timer(std::chrono::milliseconds((int)(1000.0/publish_frequency_)),
                                         std::bind(&NumberPublisherNode::publishNumber, this));
        RCLCPP_INFO(this->get_logger(), "Number Publisher has been started.");
    }

private:
    void publishNumber()
    {
        auto num = std_msgs::msg::Int64();
        num.data = number_;
        publisher_->publish(num);
    }

    // We have to declare the publisher before creating
    int number_;
    rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}